### PR TITLE
Bluetooth: HCI SPI: Explain how to connect a host to `hci_spi`

### PR DESCRIPTION
**Pull request by jori-nordic**
This explains how to connect the Zephyr host to another board running
the `hci_spi` sample.

The instructions are tailored for an nRF52840dk board, but are easily
adapted to any other board.

### CODE REVIEW

File: samples/bluetooth/hci_spi/README.rst [hunk 1]
```diff
@@ -38,3 +38,107 @@ Zephyr's Bluetooth HCI driver core; see the help associated with the
...
 to :ref:`96b_carbon_nrf51_bluetooth` for instructions specific to the
 96Boards Carbon board.
+
+Using the controller with the Zephyr host
+=========================================
+
+This describes how to hook up a board running this sample to a board running
+an application that uses the Zephyr host.
+Here the example is with two nRF52840dk boards.
+All wires are connected to the same port/pin on the other side, save for the reset line, which is connected to a GPIO on the host and the SoC reset line on the controller.
```
Comment (jhedberg) [inline samples/bluetooth/hci_spi/README.rst:48]: Could you reformat this a bit, since it seems highly inconsistent with the line lengths?

Comment (jori-nordic) [reply samples/bluetooth/hci_spi/README.rst:48]: It's to help with the reading. I have trouble with thick walls of text. I'll add a newline and reflow if that's ok

File: subsys/bluetooth/host/Kconfig [hunk 2]
```diff
@@ -47,7 +47,7 @@ config BT_HCI_TX_STACK_SIZE
 	prompt "HCI Tx thread stack size" if BT_HCI_TX_STACK_SIZE_WITH_PROMPT
 	default 512 if BT_H4
 	default 512 if BT_H5
-	default 416 if BT_SPI
+	default 1024 if BT_SPI
```
Comment (msmttchr) [inline subsys/bluetooth/host/Kconfig:50]: This looks like a significant increase in stack size, any idea on the root cause?
Is it possible that application was compiled in debug mode or with some logging enabled?

Comment (jori-nordic) [reply subsys/bluetooth/host/Kconfig:50]: I didn't really measure it. I can try to, yeah. This is probably due to some update to the nordic hal that requires more stack.

In general it's probably a bad idea to have a hand-tuned TX stack size, as the TX stack calls directly into the driver and the stack depth required by the vendor hal can vary a lot.
 
I would rather prefer comfortable defaults, like the rest of the BT stacks, and the application developer can do the fine-tuning. Nothing worse IMO than booting a sample and being greeted by the panic handler.

**Comment by jori-nordic** [comment]
> a regular microcontroller and a radio chip

It is more realistic, yes. Unfortunately, we don't make regular MCUs at nordicsemi, so we don't have any readily available for testing (eg when there are PRs to the SPI driver/sample). I don't have the carbon board either.

 I added the readme part because it's always a pain to figure out how both dts are supposed to look like.

File: subsys/bluetooth/host/Kconfig (hunk 2 above)
Comment (jori-nordic) [reply subsys/bluetooth/host/Kconfig:50]: > with some logging enabled

If we enable logging and it makes the application stack-overflow then it's a bad default IMO.

File: samples/bluetooth/hci_spi/README.rst [hunk 3]
```diff
@@ -38,3 +38,111 @@ Zephyr's Bluetooth HCI driver core; see the help associated with the
...
+
+Here the example is with two nRF52840dk boards. All wires are connected to the
+same port/pin on the other side, save for the reset line, which is connected to
+a GPIO on the host and the SoC reset line on the controller.
+
+Host board
+----------
+
+On the host application, some config options need to be used to select the SPI
+driver instead of the built-in controller:
```
Comment (jori-nordic) [inline samples/bluetooth/hci_spi/README.rst:56]: TODO: add links to relevant how-to dts and kconfig sections in doc

File: subsys/bluetooth/host/Kconfig (hunk 2 above)
Comment (msmttchr) [reply subsys/bluetooth/host/Kconfig:50]: I agree, but I experienced this in the past. I am not sure what is the recommended way to define thread's stack size.
Normally, assuming no recursive functions are present, it should be the maximum of each possible calling tree.
If no function pointers are present, I guess the compiler could do the job.
In general, it's a difficult task, but increasing all the threads stack to be on the safe side could lead to significant waste of memory.

I know also that in shell, you can inspect stack usage at runtime.

Is there a guideline in Zephyr on how to define stack size for threads?

Comment (jori-nordic) [reply subsys/bluetooth/host/Kconfig:50]: > If no function pointers are present

Doesn't zephyr use this a whole lot? at least for HW drivers definitely. I think also for HCI drivers.
I asked during today's zephyr host meeting and didn't get concrete answers, so I agree a line in the doc would be great.

**Comment by jori-nordic** [comment]
> I added the readme part because it's always a pain to figure out how both dts are supposed to look like

I'll make a blog post instead

**Comment by jori-nordic** [comment]
> I'll make a blog post instead

https://jonathan.rico.live/blog/2024-01-12/

**Pull request closed without merging**